### ISSUE TITLE
twisted.web.client.Agent should use trustRoot

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7042]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Alex reported|
|-|-|
|Trac ID|trac#7042|
|Type|defect|
|Created|2014-03-19 18:18:21Z|
|Branch|https://github.com/twisted/twisted/tree/agent-trustRoot-7042|

Right now certificates aren't actually checked to see if they are trusted by the platform, this is a significant MITM attack vector.

Attachments:

* [t7042.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b1e/b1e381586d15bf3c7ced0b62873cf2488df68991/95cb74e1d1b2c05e137c8dfe3e98eea1eeb092ab.diff) (843 bytes) - added by Alex on 2014-03-19 18:18:38Z - 
* [t7042.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b1e/b1e381586d15bf3c7ced0b62873cf2488df68991/f3edd1d6eca612d45fbc02145f51cb06f0ed8c62.diff) (1159 bytes) - added by Alex on 2014-03-19 18:29:00Z - The attached patch adds a topfile.
* [t7042.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b1e/b1e381586d15bf3c7ced0b62873cf2488df68991/8c8afa0c9a572fcd6bd9d48fd69aeff3d53d8abb.diff) (2455 bytes) - added by Alex on 2014-03-19 18:47:09Z - Adds a test.
* [t7042-platformTrust.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b1e/b1e381586d15bf3c7ced0b62873cf2488df68991/e48f06c0eb3f97940c4c55c7b977af1d6b83acf6.diff) (755 bytes) - added by Alex on 2014-03-20 16:31:44Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7042 7042
type__defect defect
reporter__Alex Alex
priority__highest highest
milestone__ 
branch__branches_agent_trustRoot_7042 branches/agent-trustRoot-7042
branch_author__dreid dreid
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1395253101000000 1395253101000000
changetime__1395684934000000 1395684934000000
version__None None
owner__dreid dreid
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to Alex|
|-|-|

Thanks.  All changes need test coverage.  Also please add a news fragment.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|Alex commented|
|-|-|

I've added a topfile, it's not clear to me what to test here.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

Without [[#4888](https://github.com/twisted/twisted/issues/4888)](https://github.com/twisted/twisted/issues/4888) this is only half-useful, which is to say adds no security. But #4888 will do the other half, and hopefully soon. So it's a good step.

Also, this is not backwards compatible. Someone is likely using Agent with the assumption that it works with any random HTTPS server. Alternatives:

1. Waive backwards compat policy in this case, given it's a significant security issue for most users.
2. Add a `SecureAgent` that does this plus hostname verification, changing all docs to use that. Not sure what to do for HTTPS usage with Agent... a warning if you don't explicitly pass in your own `CertificateOptions` instance?

Even if backwards incompat is not an issue I worry that since first option is that since it encourages developers to just continue to use Agent as is and trust it will DTRT re security, users with old Twisted installed will end up with insecure usage if developer forgets to update their requirements.txt. So I tend towards option two, where users are more likely to end up being upgraded to newer Twisted so that the new API `SecureAgent` is available.

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|Alex commented|
|-|-|

I agree that without [#4888](https://github.com/twisted/twisted/issues/4888) the validation is incomplete, and still vulnerable to MITM attacks, however this also presents a significant improvement over the status quo, and can be accomplished seperately.

I strongly advocate for the position that: this falls under the security exception to the backwards compat policy.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_9"></a>|@dreid commented|
|-|-|

I don't think there is any evidence to suggest that people are more likely to upgrade Twisted _and_ change their code vs just upgrading Twisted.

People who do not want to certificate verification are the exception not the rule.  For them we can file another ticket to add an ```InsecureWebClientContextFactoryDoNotUse```.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

> So I tend towards option two, where users are more likely to end up being upgraded to newer Twisted so that the new API `SecureAgent` is available.

Apart from the other considerations, `Foo` and `PositiveAdjectiveFoo` is a terrible naming convention and a terrible disservice to application developers.  This will just encourage people to keep using `Agent` because they don't know any better and their code will be insecure whether it is run against an older version of Twisted or a newer version.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> People who do not want to certificate verification are the exception not the rule.

We're also not talking about random people here.  We're talking about people who want an HTTPS client - not SMTP/TLS, not Tor, not anything else.

It is quite reasonable to enforce the norms for HTTPS clients in this API.  The norm for HTTPS clients is to do certificate verification.

> For them we can file another ticket to add an `InsecureWebClientContextFactoryDoNotUse`.

Or they can write it themselves.  It's not hard.  It's doing all the proper security checks that's hard. :P

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_12"></a>|@hynek commented|
|-|-|

I feel like the introduction of `InsecureWebClientContextFactoryDoNotUse` should be part of _this_ ticket.  If we potentially break something, the fix should be part of it.

Also -1 on SecureAgent; I make fun of `yaml.safe_load` all the time, let’s not introduce this abomination in Twisted.  The end-goal has to be an Agent that behaves sanely.  Either now with the fig leaf of security or later by starting a deprecation cycle.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_13"></a>|@dreid commented|
|-|-|

(In [41950]) Branching to agent-trustRoot-7042.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_14"></a>|@dreid commented|
|-|-|

Replying to [hynek](#note_12):
> I feel like the introduction of `InsecureWebClientContextFactoryDoNotUse` should be part of _this_ ticket.  If we potentially break something, the fix should be part of it.

I think that is out of scope and will only further delay an obviously positive change.

It's also not clear to me at all that twisted should provide a mechanism like that for disabling security when using HTTPS.

I can't review this because I helped with the authorship, but I've made a branch and forced the build so it should be easier for someone else to review.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_15"></a>|@itamarst commented|
|-|-|

I'll try to finish this today, if not someone else should, but:

 1. News file should note this is still insecure until [#4888](https://github.com/twisted/twisted/issues/4888) is in (we can remove this from release notes if that's merged before next release).
 2. News file should prominently note this change is not backwards compatible.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_16"></a>|@dreid commented|
|-|-|

Replying to [itamar](#note_15):
> I'll try to finish this today, if not someone else should, but:
> 
>  1. News file should note this is still insecure until [#4888](https://github.com/twisted/twisted/issues/4888) is in (we can remove this from release notes if that's merged before next release).
>  2. News file should prominently note this change is not backwards compatible.

 The news file has been updated to address this.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_17"></a>|@itamarst commented|
|-|-|

Shouldn't this be using `platformTrust()`, in order to be future compatible?

**Comment by twisted-trac** [comment]
|<img alt="Alex's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|Alex commented|
|-|-|

itamar: You're totally right, it should have been. Uploaded a patch against the branch to do so.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph set owner to @dreid|
|-|-|

Please land.

Maybe the NEWS file should say "TLS error"?  Except the exception type is SSLError, everything is terrible, never mind.

Looking forward to hostname verification.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_20"></a>|@dreid set status to `closed`|
|-|-|

(In [41986]) Merge agent-trustRoot-7042: Agent uses platformTrust by default.

Author: alex, dreid, radix

Reviewer: glyph, itamar, exarkun

Fixes: [#7042](https://github.com/twisted/twisted/issues/7042)

twisted.web.client.Agent now verifies that the provided server certificate in a TLS connection is trusted by the platform. Until [#4888](https://github.com/twisted/twisted/issues/4888) is completed there are still known MITM attack vectors against Agent. Applications which were relying on the absence of certificate checking will now receive an SSL error.
